lib/html5ever/precompiled.ex

```diff
@@ -340,4 +465,121 @@ defmodule Html5ever.Precompiled do
         {:error, "couldn't fetch NIF from #{url}: #{inspect(other)}"}
     end
   end
+
+  @doc """
+  Download a list of files from URLs and calculate its checksum.
+
+  Returns a list with details of the download and the checksum of each file.
+  """
+  def download_nif_artifacts_with_checksums!(urls) do
+    tasks =
+      Task.async_stream(urls, fn url -> {url, download_nif_artifact(url)} end, timeout: :infinity)
+
+    cache_dir = cache_dir("precompiled_nifs")
+    :ok = File.mkdir_p(cache_dir)
+
+    Enum.map(tasks, fn {:ok, result} ->
+      with {:download, {url, download_result}} <- {:download, result},
+           {:download_result, {:ok, body}} <- {:download_result, download_result},
+           hash <- :crypto.hash(@checksum_algo, body),
+           path <- Path.join(cache_dir, basename_from_url(url)),
+           {:file, :ok} <- {:file, File.write(path, body)} do
+        checksum = Base.encode16(hash, case: :lower)
+
+        Logger.debug(
+          "NIF cached at #{path} with checksum #{inspect(checksum)} (#{@checksum_algo})"
+        )
+
+        %{
+          url: url,
+          path: path,
+          checksum: checksum,
+          checksum_algo: @checksum_algo
+        }
+      else
+        {context, result} ->
+          raise "could not finish the download of NIF artifacts. Context: #{inspect(context)}. Reason: #{inspect(result)}"
+      end
+    end)
+  end
+
+  defp basename_from_url(url) do
+    uri = URI.parse(url)
+
+    uri.path
+    |> String.split("/")
+    |> List.last()
+  end
+
+  def read_map_from_file(file) do
+    with {:ok, contents} <- File.read(file),
+         {%{} = contents, _} <- Code.eval_string(contents) do
+      contents
+    else
+      _ -> %{}
+    end
+  end
+
+  # TODO: consider acquiring a lock for that file maybe with another tmp file.
+  defp write_metadata(metadata) do
```

Comment:
I wonder if we should perhaps version the file or the modules within the file, something like `{:v1, map}`.

Reply:
With the addition of the module name to the metadata file I believe we don't need to worry about this, because the write to that file will be serial (only the module of the same name will write to it).